.github/workflows/test.yml

```diff
@@ -0,0 +1,37 @@
+name: test
+# based on https://jacobian.org/til/github-actions-poetry/
+
+on:
+  pull_request:
+    types: [opened, edited, reopened, synchronize]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.11.1
+      - name: cache poetry install
+        uses: actions/cache@v2
+        with:
+          path: ~/.local
+          key: poetry-1.7.1-0
+      - uses: snok/install-poetry@v1
+        with:
+          version: 1.7.1
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+      - name: cache deps
+        id: cache-deps
+        uses: actions/cache@v2
+        with:
+          path: .venv
+          key: pydeps-${{ hashFiles('**/poetry.lock') }}
+      - run: poetry install --no-interaction --no-root
+        if: steps.cache-deps.outputs.cache-hit != 'true'
+      - run: poetry install --no-interaction
+      - run: poetry run pytest
+      - name: coverage
+        run: poetry run pytest --cov
```

Comment:
looks like this is running the tests twice, so you can remove the above step

Reply:
one's to generate a code coverage report; I thought it might be useful to have that as an extra step with the first test call to be more friendly to debugging. but maybe one call is enough. Ill remove the first one